src/fidesops/ops/service/connectors/saas/authenticated_client.py

```diff
@@ -75,27 +81,77 @@ def get_authenticated_request(
         # otherwise just return the prepared request
         return req
 
+    def retry_send(  # type: ignore
+        retry_count: int,
+        backoff_factor: float,
+        retry_status_codes: List[int] = [429, 502, 503, 504],
+    ) -> Callable:
+        """
+        Retry decorator for http requests, backing off exponentially or listening to server retry-after header
+
+        Exponential backoff factor uses the following formula:
+        backoff_factor * (2 ** (retry_attempt))
+        For an backoff_factor of 1 it will sleep for 2,4,8 seconds
+
+        General exceptions are always retried. RequestFailureResponseException exceptions are only retried
+        if the status code is in retry_status_codes.
+        """
+
+        def decorator(func: Callable) -> Callable:
+            @wraps(func)
+            def result(*args: Any, **kwargs: Any) -> Response:
+                self = args[0]
+                last_exception: Optional[Union[BaseException, Exception]] = None
+
+                for attempt in range(retry_count + 1):
+                    sleep_time = backoff_factor * (2 ** (attempt + 1))
+                    try:
+                        return func(*args, **kwargs)
+                    except RequestFailureResponseException as exc:  # pylint: disable=W0703
+                        response: Response = exc.response
+                        status_code: int = response.status_code
+                        last_exception = ClientUnsuccessfulException(
+                            status_code=status_code
+                        )
+
+                        if status_code not in retry_status_codes:
+                            break
+
+                        # override sleep time if retry after header is found
+                        retry_after_time = get_retry_after(response)
+                        sleep_time = (
+                            retry_after_time if retry_after_time else sleep_time
+                        )
+                    except Exception as exc:  # pylint: disable=W0703
+                        dev_mode_log = f" with error: {exc}" if config.dev_mode else ""
+                        last_exception = ConnectionException(
+                            f"Operational Error connecting to '{self.key}'{dev_mode_log}"
+                        )
+
+                    if attempt < retry_count:
+                        logger.warning(
+                            "Retrying http request in %s seconds", sleep_time
+                        )
+                        sleep(sleep_time)
+
+                raise last_exception  # type: ignore
+
+            return result
+
+        return decorator
+
+    @retry_send(retry_count=3, backoff_factor=1.0)  # pylint: disable=E1124
```

Comment:
maybe i'm missing something - but i was expecting these values to be config driven, rather than hardcoded here.

i assume these are just some defaults and we do plan to expose some config to drive the values here, as different endpoints and use cases have different requirements and allowances for retries. in that case, i wonder if it'd be a bit cleaner to put the default arg values in the `retry_send` method definition itself?

Reply:
We discussed offline. For now we'll go forward with these values to only be configurable within our code. This only controls retries for certain error responses, we might need to expand this in the future so that a user could customize this.